examples/android/routeguide/app/src/main/java/io/grpc/routeguideexample/RouteGuideActivity.java

```diff
@@ -0,0 +1,370 @@
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                doListFeatures(400000000, -750000000, 420000000, -730000000);
+            }
+        }).execute();
+    }
+
+    public void recordRoute(View view) {
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                List<Point> points = new ArrayList<Point>();
+                points.add(Point.newBuilder()
+                        .setLatitude(407838351).setLongitude(-746143763).build());
+                points.add(Point.newBuilder()
+                        .setLatitude(408122808).setLongitude(-743999179).build());
+                points.add(Point.newBuilder()
+                        .setLatitude(413628156).setLongitude(-749015468).build());
+                doRecordRoute(points, 5);
+            }
+        }).execute();
+    }
+
+    public void routeChat(View view) {
+        disableButtons();
+        new GrpcTask(new GrpcRunnable() {
+            @Override
+            public void run() throws Exception {
+                doRouteChat();
+            }
+        }).execute();
+    }
+
+    private void disableButtons() {
+        ((InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE))
+                .hideSoftInputFromWindow(mHostEdit.getWindowToken(), 0);
+        mGetFeatureButton.setEnabled(false);
+        mListFeaturesButton.setEnabled(false);
+        mRecordRouteButton.setEnabled(false);
+        mRouteChatButton.setEnabled(false);
+    }
+
+    private class GrpcTask extends AsyncTask<Void, Void, String> {
+        private String mHost;
+        private String mMessage;
+        private int mPort;
+        private final GrpcRunnable mGrpc;
+    
+        GrpcTask(GrpcRunnable grpc) {
+            this.mGrpc = grpc;
+        }
+
+        @Override
+        protected void onPreExecute() {
+            mHost = mHostEdit.getText().toString();
+            String portStr = mPortEdit.getText().toString();
+            mPort = TextUtils.isEmpty(portStr) ? 0 : Integer.valueOf(portStr);
+            mResultText.setText("");
+        }
+
+        @Override
+        protected String doInBackground(Void... nothing) {
+            try {
+                mChannel = ManagedChannelBuilder.forAddress(mHost, mPort)
+                        .usePlaintext(true)
+                        .build();
+                mBlockingStub = RouteGuideGrpc.newBlockingStub(mChannel);
+                mAsyncStub = RouteGuideGrpc.newStub(mChannel);
+                mGrpc.run();
+                return "Success!";
+            } catch (Exception e) {
+                return "Failed... : " + e.getMessage();
+            }
+        }
+
+        @Override
+        protected void onPostExecute(String result) {
+            try {
+                mChannel.shutdown().awaitTermination(1, TimeUnit.SECONDS);
```

Comment:
So you wait until the channel shuts down before updating the UI? Maybe setting `idleTimeout` on `ManagedChannelBuilder` and avoiding shutdown here would be better? Although it still would create multiple channels. Maybe only initialize channel if it is null?